Feat/stackerdb chunk db

jcnelson opened the pull request:
This implements the Stacker DB chunk storage database, as part of breaking apart #3534.

kantai commented on src/net/stackerdb/mod.rs:
```diff
@@ -131,3 +139,66 @@ pub struct StackerDBSyncResult {
+/// Chunk metadata from the DB.
+/// This is derived state from a StackerDBChunkData message.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ChunkMetadata {
```
This struct and its fields should be referenced in the `StackerDB` rustdocs above -- where you currently provide an overview of what the StackerDB is and how it works, it should contain references to where the different items are implemented.

kantai commented on src/net/stackerdb/db.rs:
```diff
@@ -0,0 +1,602 @@
+    format!("stackerdb_{}", normalized_name)
+}
+
+/// Load up chunk metadata from the database, given the primary key.
```
```suggestion
/// Load a chunk's metadata from the database, keyed by the chunk's smart contract, the reward cycle consensus hash, and the chunk identifier.
```

Looking at this PR after #3551 and #3552, I think there's a naming conflict with "StackerDB" -- you refer to the local database as "the `StackerDB`", but also, throughout the code, different smart contracts' associated chunk storage is each called "a `StackerDB`". This is already kind of confusing to me, and this is pretty early on in the life of this code, so I think we'd probably be better served with a different way to talk about these things.

jcnelson replied:
I'm having a hard time understanding what the ask here is.  The "smart contracts' associated chunk storage" _is_ a stacker DB.  The node maintains a local replica of that state for each stacker DB it subscribes to.  But, I don't see the value in calling the local replica by a different name, since I think that's already understood simply because a single process cannot be a whole p2p network.

jcnelson replied:
I think you're referring to the fact that `StackerDB` the struct currently stores all stacker DB chunks for all subscribed stacker DBs?  If so, then I'll rename `StackerDB` to `StackerDBSet` to disambiguate it.

kantai commented on src/net/stackerdb/db.rs:
```diff
@@ -0,0 +1,602 @@
+        query_rows(&self.conn(), sql, NO_PARAMS).map_err(|e| e.into())
+    }
+
+    /// Set up a database's storage slots.
+    /// The slots must be in a deterministic order, since they are used to determine the chunk ID
+    /// (and thus the key used to authenticate them)
```
What is a storage slot? It seems like this method is just writing empty data into the table. Why is this necessary?

jcnelson replied:
It's done this way because the follow-on StackerDB state sync PR needs to be able to extract the version vector for all of the DB's slots.  But in order to do that, those slots either have to be instantiated, or we need some kind of sparse version vector calculation.  The latter requires one SQL query per slot (since if slot data does not exist, we'd need to synthesize it on the fly), whereas this only requires a single query which can use an index.

kantai commented on src/net/stackerdb/db.rs:
```diff
@@ -0,0 +1,602 @@
+    }
+
+    /// Insert a chunk into the DB.
+    /// It must be authenticated, and its lamport clock must be higher than the one that's already
+    /// there.  These will not be checked.
```
```suggestion
    /// The caller must check that this is authenticated and that its lamport clock is higher than the existing one. These are not checked by this method.
```

kantai reviewed (changes requested):
I have a couple of comments and requests that should be addressed:

1. Dynamic table names are pretty close to a sql anti-pattern and can lead to issues down the road (as well as requiring contract name sanitization), and so I think unless there's a very compelling reason to do this, it should be avoided.
2. `StackerDB` refers to two different things in the codebase (and the docs) and I think it's making the codebase confusing (sometimes "stackerdb" is a contract identifier, other times its the database storing multiple "stackerdb"s), and there's no better time than now to try to fix this (if it isn't fixed now, the codebase will probably have to live with this confusion forever)

kantai commented on src/net/stackerdb/bits.rs:
```diff
@@ -0,0 +1,123 @@
+impl ChunkMetadata {
+    /// Get the digest to sign that authenticates this chunk data and metadata
+    fn sighash(&self) -> Sha512Trunc256Sum {
+        let mut bytes = vec![];
+        bytes.extend_from_slice(&self.rc_consensus_hash.0);
+        bytes.extend_from_slice(&self.chunk_id.to_be_bytes());
+        bytes.extend_from_slice(&self.chunk_version.to_be_bytes());
+        bytes.extend_from_slice(&self.data_hash.0);
+
+        Sha512Trunc256Sum::from_data(&bytes)
```
Avoid a copy:

```suggestion
        let mut hasher = Sha512_256::new();
        hasher.update(&self.rc_consensus_hash.0);
        hasher.update(&self.chunk_id.to_be_bytes());
        hasher.update(&self.chunk_version.to_be_bytes());
        hasher.update(&self.data_hash.0);

        Sha512Trunc256Sum::from_hasher(hasher)
```

jcnelson replied:
Yup, will do

kantai commented on src/net/stackerdb/bits.rs:
```diff
@@ -0,0 +1,123 @@
+        Sha512Trunc256Sum::from_data(&bytes)
+    }
+
+    /// Sign this chunk metadata, committing to rc_consensus_hash, chunk_id, chunk_version, and
+    /// data_hash.
```
Can you add a note to the rustdoc that the signature will be stored in `self.signature`?

jcnelson replied:
Yup

kantai commented on src/net/stackerdb/bits.rs:
```diff
@@ -0,0 +1,123 @@
+        }
+    }
+
+    /// Sign this given chunk data message with the given private key.  Binds it to a particular
+    /// reward cycle in a particular PoX history (identified by `rc_consensus_hash`).
+    /// Fails if the underlying signing library fails.
```
Can you add a note to the rustdoc that the signature will be stored in `self.signature`?

jcnelson replied:
Yup

kantai commented on src/net/stackerdb/tests/db.rs:
```diff
@@ -0,0 +1,580 @@
+/// Test that we can instantiate a stacker DB instance
+#[test]
+fn test_stackerdb_connect() {
+    let path = "/tmp/test_stackerdb_connect.sqlite";
```
Can you prefix these paths with `/tmp/stacks-node-tests/stackerdb/`? It makes cleaning up `/tmp` much easier after running the test suite.

jcnelson replied:
Sure

kantai reviewed (commented):
Just a few more comments over my last review, which I think is still valid -- mostly, I think that programmatic table construction should be avoided unless there is a really compelling reason.

jcnelson commented on src/net/stackerdb/bits.rs:
```diff
@@ -0,0 +1,123 @@
+
+impl ChunkMetadata {
+    /// Get the digest to sign that authenticates this chunk data and metadata
+    fn sighash(&self) -> Sha512Trunc256Sum {
```
Yup

jcnelson commented on src/net/stackerdb/db.rs:
```diff
@@ -0,0 +1,602 @@
+        .replace(".", "_");
+
+    format!("stackerdb_{}", normalized_name)
+}
```
I removed the one-table-per-StackerDB code and now just store chunks in a single table.

jcnelson commented on src/net/stackerdb/bits.rs:
```diff
@@ -0,0 +1,123 @@
+
+/// This module contains methods for interacting with the data contained within the messages
+use crate::net::{
+    Error as net_error, StackerDBChunkData, StackerDBChunkInvData, StackerDBGetChunkData,
```
This and the other `StackerDB*` structs are new protocol messages, whose definitions IMHO belong with the other message definitions.

jcnelson commented on src/net/stackerdb/mod.rs:
```diff
@@ -131,3 +139,66 @@ pub struct StackerDBSyncResult {
+/// Chunk metadata from the DB.
+/// This is derived state from a StackerDBChunkData message.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ChunkMetadata {
```
Done

obycode reviewed (approved):
LGTM!

jcnelson commented:
Hey @kantai, this is ready for re-review when you get the chance.  Thanks!

kantai commented on stackslib/src/net/mod.rs:
```diff
@@ -256,6 +256,20 @@ pub enum Error {
     ExpectedEndOfStream,
     /// burnchain error
     BurnchainError(burnchain_error),
+    /// chunk is stale
+    StaleChunk(u32, u32),
```
```suggestion
    /// chunk is stale
    StaleChunk { supplied_version: u32, latest_version: u32 },
```

kantai commented on stackslib/src/net/mod.rs:
```diff
@@ -256,6 +256,20 @@ pub enum Error {
+    /// slot signer is wrong
+    BadSlotSigner(StacksAddress, u32),
+    /// too many writes to a slot
+    TooManySlotWrites(u32, u32),
```
```suggestion
    TooManySlotWrites { supplied_version: u32, max_writes: u32 },
```

kantai commented on stackslib/src/net/stackerdb/db.rs:
```diff
@@ -0,0 +1,554 @@
+        let open_flags = if path != ":memory:" {
+            match fs::metadata(path) {
+                Err(e) => {
+                    if e.kind() == io::ErrorKind::NotFound {
+                        // need to create
+                        if readwrite {
+                            create_flag = true;
+                            let ppath = Path::new(path);
+                            let pparent_path = ppath
+                                .parent()
+                                .expect(&format!("BUG: no parent of '{}'", path));
+                            fs::create_dir_all(&pparent_path).map_err(|e| db_error::IOError(e))?;
+
+                            OpenFlags::SQLITE_OPEN_READ_WRITE | OpenFlags::SQLITE_OPEN_CREATE
+                        } else {
+                            return Err(db_error::NoDBError.into());
+                        }
+                    } else {
+                        return Err(db_error::IOError(e).into());
+                    }
```
```suggestion
                    if e.kind() != io::ErrorKind::NotFound {
                        return Err(db_error::IOError(e).into());
                    }
                    // path not found, must create db path.
                    if !readwrite {
                        return Err(db_error::NoDBError.into());
                    }
                    create_flag = true;
                    let ppath = Path::new(path);
                    let pparent_path = ppath
                                .parent()
                                .expect(&format!("BUG: no parent of '{}'", path));
                    fs::create_dir_all(&pparent_path).map_err(|e| db_error::IOError(e))?;

                    OpenFlags::SQLITE_OPEN_READ_WRITE | OpenFlags::SQLITE_OPEN_CREATE
```

kantai commented on stackslib/src/net/stackerdb/db.rs:
```diff
@@ -0,0 +1,554 @@
+        let qry = "SELECT data FROM chunks WHERE stackerdb_id = ?1 AND slot_id = ?2";
+        let args: &[&dyn ToSql] = &[&stackerdb_id, &slot_id];
+
+        let mut stmt = self
+            .conn
+            .prepare(&qry)
+            .map_err(|e| net_error::DBError(e.into()))?;
+
+        let mut rows = stmt.query(args).map_err(|e| net_error::DBError(e.into()))?;
+
+        let mut data = None;
+        while let Some(row) = rows.next().map_err(|e| net_error::DBError(e.into()))? {
+            data = Some(row.get_unwrap(0));
+            break;
+        }
+        Ok(data)
```
You can get a single row result from a database, wrapped in an option by using `query_row().optional()`. That replaces lines 526-538 with just:

```suggestion
     self.conn.query_row(qry, args, |row| row.get(0)).optional().map_err(|e| e.into())
```

kantai reviewed (commented):
This looks good to me, my remaining feedback is pretty superficial.

jcnelson commented:
@kantai I addressed all your feedback in `feat/stackerdb-rpc` (by mistake -- was using the wrong tab).  Is that okay?  It'll find its way into this branch.

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
